packages/realm/src/Object.ts

```diff
@@ -185,7 +185,11 @@ export class RealmObject<T = DefaultObject> {
     const table = binding.Helpers.getTable(realm.internal, tableKey);
     if (primaryKey) {
       const primaryKeyHelpers = properties.get(primaryKey);
-      const primaryKeyValue = values[primaryKey];
+      const defaultValue = primaryKeyHelpers.default;
+      let primaryKeyValue = values[primaryKey];
+      if (!primaryKeyValue && typeof defaultValue !== "undefined") {
```

Comment:
I'd say this would render incorrect behaviour if called with a falsy value (`0`, `false`, `""`, etc.), in which case the default should not override.

We could either replace this with a check for undefined (which I think is the current behaviour on v11):

```suggestion
      if (primaryKeyValue === undefined && typeof defaultValue !== "undefined") {
```

Or alternatively a check to see if the value was provided in the values (as an enumerable property) - but that's probably a bit less flexible on callers:

```suggestion
      if (!(primaryKey in values) && typeof defaultValue !== "undefined") {
```